### ISSUE TITLE
Got "TypeError: must be convertible to a buffer, not bytearray" in PyPy

### ISSUE THREAD

**Issue by ethe** [andymccurdy/redis-py#708]
```
PyPy version: 4.0.1 with GCC 4.8.4 on linux2
System: Ubuntu 15.04 x86 64bit
Error line: ./site-packages/redis/connection.py line 359 self._reader.feed(self._buffer, 0, bufflen)
```

``` pytb
Traceback (most recent call last):
  File "/home/zeno/amap/env/pypy/site-packages/redis/client.py", line 572, in execute_command
    connection.send_command(*args)
  File "/home/zeno/amap/env/pypy/site-packages/redis/connection.py", line 564, in send_command
    self.send_packed_command(self.pack_command(*args))
  File "/home/zeno/amap/env/pypy/site-packages/redis/connection.py", line 539, in send_packed_command
    self.connect()
  File "/home/zeno/amap/env/pypy/site-packages/redis/connection.py", line 447, in connect
    self.on_connect()
  File "/home/zeno/amap/env/pypy/site-packages/redis/connection.py", line 515, in on_connect
    if nativestr(self.read_response()) != 'OK':
  File "/home/zeno/amap/env/pypy/site-packages/redis/connection.py", line 578, in read_response
    response = self._parser.read_response()
  File "/home/zeno/amap/env/pypy/site-packages/redis/connection.py", line 360, in read_response
    self._reader.feed(self._buffer, 0, bufflen)
TypeError: must be convertible to a buffer, not bytearray

```

Now I just move hiredis out, redis clients works no problem.

**Comment by ethe** [comment]
I checked [hiredis-py](https://github.com/redis/hiredis-py)'s usage, method `feed` should input a string, instead of a bytearray.

**Comment by ethe** [comment]
This error was catched so can't be throwed out, only use single-step debug can find it.

**Comment by andymccurdy** [comment]
I'm not sure about this. The whole point of using a `bytearray` is to prevent memory allocations. Support for `bytearray` was specifically added to hiredis-py here: https://github.com/redis/hiredis-py/pull/22

Casting the `bytearray` to `bytes` creates an additional memory allocation every call. I'd really like to avoid that.

Does PyPy treat `bytearray` objects differently than CPython does? From the Python docs, the two should be completely interchangeable:

"Both bytes and bytearray objects support the common sequence operations. They interoperate not just with operands of the same type, but with any bytes-like object. Due to this flexibility, they can be freely mixed in operations without causing errors."
https://docs.python.org/3/library/stdtypes.html#bytes-methods

**Comment by ethe** [comment]
Yeah, I'm sure that the behavior of bytearray is not different on pypy 4.0.1. Maybe I provide a simple way to reproduce it? I think could just judge if interpreter is pypy then use bytes method?

**Comment by ethe** [comment]
Can simply reproduce by below program.
```python
import hiredis


test = bytearray()

reader = hiredis.Reader()
reader.feed(test)
```
Python 2.7.10 run it no problem, but use PyPy 4.0.1 got below error.
```
 ➜  pypy reproduce.py              
Traceback (most recent call last):
  File "reproduce.py", line 7, in <module>
    reader.feed(test)
TypeError: must be convertible to a buffer, not bytearray
```

**Comment by andymccurdy** [comment]
It looks like this is a hiredis-py error. The `bytearray` test in the hiredis-py suite (https://github.com/redis/hiredis-py/blob/master/test/reader.py#L168) fails when running against PyPy 4.0.1 raising the error you're seeing.

Also, these two tests (https://github.com/redis/hiredis-py/blob/master/test/reader.py#L133-L139) cause the test runner to segfault against PyPy 4.0.1. Not sure why, but this is likely another problem you'll run into trying to run hiredis-py on PyPy. I commented them out temporarily to get the rest of the sutie to run and see the `bytearray` test error.

I'm not very familiar with PyPy (certainly not at the C extension level), so I'm not sure where to really start.

**Comment by ethe** [comment]
Well, I think the temporary solution to solve this problem is disabling hiredis-py when use PyPy, I think pure Python code run on PyPy have an acceptable performance.

**Comment by andymccurdy** [comment]
I created an issue on the hiredis-py repo. Hopefully someone more familiar with PyPy than I can take a look.

**Comment by andymccurdy** [comment]
PyPy fixed their ByteArray support in version 5.3. hiredis-py (and redis-py with hiredis-py enabled) works with PyPy 5.3 or later.
